integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,87 @@
+package kotlinx.coroutines.experimental.androidx.lifecycle
+
+import androidx.lifecycle.GenericLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event.*
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.experimental.CoroutineScope
+import kotlinx.coroutines.experimental.Dispatchers
+import kotlinx.coroutines.experimental.Job
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Creates a [Job] that will be cancelled when the [Lifecycle] encounters [cancelEvent]
+ * (which defaults to [ON_DESTROY]).
+ *
+ * Note that [cancelEvent] supports only "**down**" events. That means that only [ON_PAUSE],
+ * [ON_STOP] and [ON_DESTROY] are supported.
+ *
+ * If the [Lifecycle] is already destroyed, then the returned [Job] comes already cancelled.
+ */
+fun Lifecycle.createJob(cancelEvent: Lifecycle.Event = ON_DESTROY): Job {
+    require(cancelEvent in allowedCancelEvents) { "$cancelEvent is forbidden for createJob(…)." }
+    return Job().also { job ->
+        if (currentState == Lifecycle.State.DESTROYED) job.cancel()
+        else addObserver(object : GenericLifecycleObserver {
+            override fun onStateChanged(source: LifecycleOwner?, event: Lifecycle.Event) {
+                if (event == cancelEvent) {
+                    removeObserver(this)
+                    job.cancel()
+                }
+            }
+        })
+    }
+}
+
+private val allowedCancelEvents = arrayOf(ON_PAUSE, ON_STOP, ON_DESTROY)
+private val lifecycleJobs = ConcurrentHashMap<Lifecycle, Job>()
```

Comment:
Why do you need this (and lifecycleCoroutineScopes) global states? Isn't would better to create new Job each time this significantly simplify code and also allows to avoid shared mutable state (so one part of the code cancels a job that required for another).
If someone really wants to share a job, he can just cache instance of job.

I think current implementation can cause bugs and not enough explicit

Reply:
When getting a property, I don't expect it to give me a new value each time, that's why I implemented caching, and it is more efficient if you call it multiple times.

Keep in mind that for scopes, the `job` is only used as a parent, so you should not cancel the scope's job yourself when it's already done for you, like here with `Lifecycle`s.

In the cases where you want your scope to be shorter lived, you can still use `createScope()` or `createJob()`, and you can use the latter to make your custom lifecycle aware scope.

About misuse, you can always write `coroutineContext[Job]?.cancel()` on any scope and possibly cause bugs. Do you think will use `lifecycle.job` and then call `cancel()` on it? Also, do you think I should make it private and expose only `coroutineScope` and the two create methods?

Reply:
Do `Lifecycle` instances have an stable identity? It seems like it compares references, and as anonymous objects they can capture state and leak.

Reply:
@pakoito Yes, `Lifecycle` instances are never reused. After it is destroyed, it will be garbage collected, unless you keep a strong reference to it. Regarding leaks, there's none because the entries are always removed from the map after the `currentState` is no longer at least the passed `activeWhile` state (and `DESTROYED` is forbidden as a parameter as it does not make sense and would create leaks anyway, both semantically and technically).

Reply:
Marking this as resolved, if you still have concerns, please, use the latest version (in the **Files changed** tab) before you make a review or add a comment.